src/lib/mina_base/account_timing.ml

```diff
@@ -307,3 +307,216 @@ let if_ b ~(then_ : var) ~(else_ : var) =
 let deriver obj =
   let open Fields_derivers_zkapps in
   iso_record ~to_record ~of_record As_record.deriver obj
+
+(** A module defining the vesting parameter update procedure from the slot
+    reduction MIP. See
+    https://github.com/MinaProtocol/MIPs/blob/main/MIPS/mip-0006-slot-reduction-90s.md.
+    To summarize, this vesting parameter update is intended to be applied as an
+    edit to the ledgers during a hard fork that reduces the slot time by half.
+    The update only applies to accounts that are actively vesting at the time of
+    the hard fork; these are accounts that either have not reached their cliff
+    time yet, or have reached their cliff time but still have a positive minimum
+    balance. The update procedure does the best it can to adjust the account
+    timing information so that actively vesting accounts will still unlock funds
+    at the same system times as they would have had the hard fork not occurred.
+    It will do this perfectly for any account with a "reasonable" vesting
+    schedule, i.e., one that will finish before half the remaining life of the
+    chain (at the time of the hard fork) has elapsed.
+
+    An important note on the parameters: the slot at which the hard fork occurs
+    is called [hardfork_slot] below. This slot is the global slot since genesis
+    that is set in the genesis constants of the new (post hard fork) chain. It
+    is not the global slot of the hard fork block, nor the [slot_tx_end], nor
+    the [slot_chain_end]. The update procedure also assumes that the genesis
+    timestamp of the new chain is set to be equal to the timestamp of the
+    [hardfork_slot] {i relative to the old (pre hard fork) chain}. This
+    simplifies the equations.
+*)
+module Slot_reduction_update = struct
+  open Unsigned
+
+  (** A form of [As_record.t] where all the fields are lifted to
+      the same [UInt64.t] type, so the arithmetic in the vesting update
+      equations becomes simpler. *)
+  type t =
```

Comment:
Nit: Checking on this code again, it seems it's not really necessary to track the data as a `As_record.t`, that will make any account need to be updated have to go through a round of `to_record` and `of_record`. Additionally, I think the initial intention of those functions are to support circuit arithmetic, so could be a misuse here. 

I think the interface would be nicer if we're tracking data as `_ Poly.t`, but for sure the implementation would be complicated a bit, just dispatch on Timed/Untimed for each function.